File: DotNet/Serie2/ConcurrentQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Serie2
{
    public class ConcurrentQueue<T> where T: class
    {
        private class Node
        {
            internal readonly T item;
            internal volatile Node next;

            internal Node(T item, Node next)
            {
                this.item = item;
                this.next = next;
            }
        }

        private readonly Node dummy;
        private volatile Node head;
        private volatile Node tail;

        public ConcurrentQueue()
        {
            dummy = new Node(null, null);
            head = dummy;
            tail = dummy;

        }

        public void Put(T t)
        {
            Node newNode = new Node(t, null);
            while (true)
            {
                Node currTail = tail;
                Node tailNext = currTail.next;
                if(currTail == tail)
                {
                    if(tailNext != null)
                    {
                        Interlocked.CompareExchange(ref tail, tailNext, currTail);
                    }
                    else
                    {
                        if(Interlocked.CompareExchange(ref currTail.next, newNode, null) == newNode)
                        {
                            Interlocked.CompareExchange(ref tail, newNode, currTail);
                            return;
                        }
                    }
                }
            }
        }

        public T TryTake()
        {
            while (true)
            {
                Node currHead = head;
                Node headNext = currHead.next;
                if(headNext == null)
                {
                    return null;
                }
                if(Interlocked.CompareExchange(ref head, headNext, currHead) == currHead)
                {
                    return currHead.item;
                }
            }
        }

        public bool IsEmpty()
        {
            return head.Equals(tail);
        }
    }
}

File: Java/Serie2/Serie2/src/ConcurrentQueue.java
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.concurrent.atomic.AtomicReference;

public class ConcurrentQueue<T> {

    private class Node<T>{
        final T item;
        final AtomicReference<Node> next;
        private Node(T item, Node next){
            this.item = item;
            this.next = new AtomicReference(next);
        }
    }

    private final Node dummy;
    private final AtomicReference<Node> head;
    private final AtomicReference<Node> tail;

    public ConcurrentQueue(){
        dummy = new Node(null, null);
        head = new AtomicReference(dummy);
        tail = new AtomicReference(dummy);
    }

    public void put(T t){
        Node newNode = new Node(t, null);
        while(true){
            Node curTail = tail.get();
            Node tailNext = (Node)curTail.next.get();
            if(curTail == tail.get()){
                if(tailNext != null){
                    tail.compareAndSet(curTail, tailNext);
                } else {
                    if(curTail.next.compareAndSet(null, newNode)){
                        tail.compareAndSet(curTail, newNode);
                        return;
                    }
                }
            }
        }
    }

    public boolean isEmpty(){
        return head.equals(tail);
    }

    public T tryTake(){
        while (true){
            Node curHead = head.get();
            Node headNext = (Node)curHead.next.get();
            if(headNext == null){
                return null;
            }
            if(head.compareAndSet(curHead, headNext)){
                return (T)curHead.item;
            }

        }
    }
}
